File: src/userspace/demo_client.rs
//! Built-in demo client for the compositor.
//!
//! Runs as a kernel thread, connects to the compositor over IPC,
//! and renders an animated gradient surface.

use crate::ipc::gui::*;
use crate::process::Pid;

const DEMO_MAX_W: u32 = 1280;
const DEMO_MAX_H: u32 = 960;
const DEMO_MIN_W: u32 = 320;
const DEMO_MIN_H: u32 = 240;
const DEMO_FRAME_MS: u64 = 16;
const DEMO_SURFACE_ID: u32 = 1;

const fn scaled_extent(available: u32, max: u32, min: u32) -> u32 {
    let target = available.saturating_mul(3) / 4;
    if available < min {
        available
    } else {
        let capped = if target < max { target } else { max };
        if capped > min {
            capped
        } else {
            min
        }
    }
}

const fn demo_surface_geometry(fb_w: u32, fb_h: u32) -> (u32, u32, i32, i32) {
    let w = scaled_extent(fb_w, DEMO_MAX_W, DEMO_MIN_W);
    let h = scaled_extent(fb_h, DEMO_MAX_H, DEMO_MIN_H);
    let x = (fb_w.saturating_sub(w) / 2) as i32;
    let y = (fb_h.saturating_sub(h) / 2) as i32;
    (w, h, x, y)
}

fn framebuffer_size() -> (u32, u32) {
    crate::drivers::virtio_gpu::get_fb_info()
        .or_else(crate::drivers::framebuffer::get_bga_fb_info)
        .map(|(_, w, h, _)| (w, h))
        .unwrap_or((1024, 768))
}
const COMPOSITOR_CHAN: u32 = 3;
unsafe fn any_as_bytes<T: Sized>(x: &T) -> &[u8] {
    unsafe { core::slice::from_raw_parts(x as *const T as *const u8, core::mem::size_of::<T>()) }
}

/// Send a message to the compositor channel.
fn send_msg(opcode: u8, payload: &[u8]) {
    let mut buf = [0u8; 256];
    buf[0] = opcode;
    if !payload.is_empty() {
        let n = payload.len().min(255);
        buf[1..1 + n].copy_from_slice(&payload[..n]);
    }
    let _ = crate::ipc::send(COMPOSITOR_CHAN, Pid(0), &buf[..1 + payload.len().min(255)]);
}

/// Kernel-thread entry for the built-in demo client.
pub fn demo_client_main(_arg: *const ()) {
    let pid = crate::process::scheduler::SCHEDULER.current_pid();
    if let Some(pid) = pid {
        crate::timer::sleep_ms(pid, 1000);
        crate::process::scheduler::yield_now();
    }

    let (fb_w, fb_h) = framebuffer_size();
    let (width, height, pos_x, pos_y) = demo_surface_geometry(fb_w, fb_h);
    // 1. Create SHM via the kernel SHM module
    let shm_id = match crate::ipc::shm::SHM
        .lock()
        .create(Pid(0), (width * height * 4) as usize)
    {
        Ok(id) => id,
        Err(_) => {
            crate::klog!(crate::klog::Level::Error, "[DemoClient] SHM create failed");
            return;
        }
    };

    let shm_addr = match crate::ipc::shm::SHM.lock().attach(shm_id, Pid(0)) {
        Ok(addr) => addr,
        Err(_) => {
            crate::klog!(crate::klog::Level::Error, "[DemoClient] SHM attach failed");
            return;
        }
    };

    crate::klog!(
        crate::klog::Level::Info,
        "[DemoClient] SHM created id={} addr=0x{:x} ({}x{} = {} bytes)",
        shm_id,
        shm_addr,
        width,
        height,
        width * height * 4,
    );

    // 2. Connect (OpCode 1)
    let conn = ConnectMsg { pid: 0 };
    send_msg(OpCode::Connect as u8, unsafe { any_as_bytes(&conn) });

    // 3. Create surface (OpCode 2)
    let surf = CreateSurfaceMsg { surface_id: DEMO_SURFACE_ID, width, height };
    send_msg(OpCode::CreateSurface as u8, unsafe { any_as_bytes(&surf) });

    // 4. BufferAttach
    let attach = BufferAttachMsg {
        surface_id: DEMO_SURFACE_ID,
        shm_id,
        width,
        height,
    };
    send_msg(OpCode::BufferAttach as u8, unsafe { any_as_bytes(&attach) });

    // 5. Set position
    let pos = SetPositionMsg {
        surface_id: DEMO_SURFACE_ID,
        x: pos_x,
        y: pos_y,
    };
    send_msg(OpCode::SetPosition as u8, unsafe { any_as_bytes(&pos) });

    let shm_ptr = shm_addr as *mut u32;
    let mut tick: u32 = 0;

    // 6. Render loop
    loop {
        tick = tick.wrapping_add(1);

        for y in 0..height {
            for x in 0..width {
                let r = (x.wrapping_add(tick)) & 0xFF;
                let g = (y.wrapping_add(tick)) & 0xFF;
                let b = (x.wrapping_add(y).wrapping_add(tick)) & 0xFF;
                unsafe {
                    let idx = (y as usize) * (width as usize) + (x as usize);
                    shm_ptr
                        .add(idx)
                        .write_volatile((r << 16) | (g << 8) | b | 0xFF000000);
                }
            }
        }

        let flush = FlushMsg {
            surface_id: DEMO_SURFACE_ID,
            x: 0,
            y: 0,
            width,
            height,
        };
        send_msg(OpCode::Flush as u8, unsafe { any_as_bytes(&flush) });

        if let Some(pid) = pid {
            crate::timer::sleep_ms(pid, DEMO_FRAME_MS);
        }
        crate::process::scheduler::yield_now();
    }
}

const _: () = {
    let bga = demo_surface_geometry(1024, 768);
    assert!(bga.0 == 768 && bga.1 == 576 && bga.2 == 128 && bga.3 == 96);

    let small = demo_surface_geometry(320, 200);
    assert!(small.0 == 320 && small.1 == 200 && small.2 == 0 && small.3 == 0);
};

File: userspace/demo_client.c
#include <stdint.h>
#include <stddef.h>

#define ZIQA_CAP_REQUEST 1000
#define ZIQA_SHM_CREATE 1010
#define ZIQA_SHM_ATTACH 1011
#define ZIQA_IPC_SEND 1021

static inline uint64_t ziqa_syscall(uint64_t nr, uint64_t arg0, uint64_t arg1, uint64_t arg2, uint64_t arg3) {
    uint64_t ret;
    register uint64_t r10 __asm__("r10") = arg3;
    __asm__ volatile (
        "syscall"
        : "=a"(ret)
        : "a"(nr), "D"(arg0), "S"(arg1), "d"(arg2), "r"(r10)
        : "rcx", "r11", "memory"
    );
    return ret;
}

// OpCodes
#define OP_CONNECT 1
#define OP_CREATE_SURFACE 2
#define OP_FLUSH 3
#define OP_BUFFER_ATTACH 5
#define OP_SET_POSITION 6

// Structures matching ipc/gui.rs
typedef struct {
    uint64_t pid;
} ConnectMsg;

typedef struct {
    uint32_t width;
    uint32_t height;
} CreateSurfaceMsg;

typedef struct {
    uint32_t surface_id;
    uint32_t x;
    uint32_t y;
    uint32_t width;
    uint32_t height;
} FlushMsg;

typedef struct {
    uint32_t surface_id;
    uint32_t shm_id;
    uint32_t width;
    uint32_t height;
} BufferAttachMsg;

typedef struct {
    uint32_t surface_id;
    int32_t x;
    int32_t y;
} SetPositionMsg;

// Send helper
static void send_msg(uint32_t chan, uint8_t opcode, const void *payload, size_t payload_len) {
    uint8_t buf[256];
    buf[0] = opcode;
    for (size_t i = 0; i < payload_len && i < 254; i++) {
        buf[1 + i] = ((const uint8_t*)payload)[i];
    }
    ziqa_syscall(ZIQA_IPC_SEND, chan, (uint64_t)buf, 1 + payload_len, 0);
}

void _start() {
    uint32_t width = 320;
    uint32_t height = 240;
    uint32_t size = width * height * 4;

    // 1. Create SHM
    uint32_t shm_id = ziqa_syscall(ZIQA_SHM_CREATE, size, 0, 0, 0);

    // 2. Attach SHM
    uint64_t shm_addr = ziqa_syscall(ZIQA_SHM_ATTACH, shm_id, 0, 0, 0);
    uint32_t *shm_ptr = (uint32_t*)shm_addr;

    uint32_t comp_chan = 3;

    // 3. Connect
    ConnectMsg conn = { .pid = 0 };
    send_msg(comp_chan, OP_CONNECT, &conn, sizeof(conn));

    // 4. Create surface
    CreateSurfaceMsg surf = { .width = width, .height = height };
    send_msg(comp_chan, OP_CREATE_SURFACE, &surf, sizeof(surf));

    // 5. BufferAttach
    BufferAttachMsg attach = { .surface_id = 1, .shm_id = shm_id, .width = width, .height = height };
    send_msg(comp_chan, OP_BUFFER_ATTACH, &attach, sizeof(attach));

    // 6. SetPosition
    SetPositionMsg pos = { .surface_id = 1, .x = 352, .y = 264 };
    send_msg(comp_chan, OP_SET_POSITION, &pos, sizeof(pos));

    uint32_t tick = 0;
    while (1) {
        tick++;
        for (uint32_t y = 0; y < height; y++) {
            for (uint32_t x = 0; x < width; x++) {
                uint32_t r = (x + tick) & 0xFF;
                uint32_t g = (y + tick) & 0xFF;
                uint32_t b = (x + y + tick) & 0xFF;
                shm_ptr[y * width + x] = (0xFF << 24) | (r << 16) | (g << 8) | b;
            }
        }

        FlushMsg flush = { .surface_id = 1, .x = 0, .y = 0, .width = width, .height = height };
        send_msg(comp_chan, OP_FLUSH, &flush, sizeof(flush));

        // Yield/nanosleep via syscall 230 (nanosleep / 16ms)
        ziqa_syscall(230, 16, 0, 0, 0);
    }
}
